### ISSUE TITLE
Add support for PostGIS' `GEOMETRY` to dialect `postgres`

### ISSUE THREAD

**Issue by xamanu** [sqlfluff/sqlfluff#1290]
Using the very common and popular [PostGIS extension](https://postgis.net/) of Postgres:

```
CREATE TABLE public.foo
(
    quadkey TEXT,
    geometry GEOMETRY (GEOMETRY, 4326),
    PRIMARY KEY (quadkey)
);
```
Executed with

```
sqlfluff lint --dialect postgres
```

causes following error:

```
L:   4 | P:  26 | L014 | Inconsistent capitalisation of unquoted identifiers.
```
It would be nice to get this added. Many thanks for the nice linter!

**Comment by WittierDinosaur** [comment]
@tunetheweb I think the ask is to include PostGIS keywords which aren't supported in regular SQL, like GEOMETRY

**Comment by tunetheweb** [comment]
OK but it's not complaining about the keyword not being available so it is parsing. The only error it's throwing on linting is the inconsistent capitalisation error (which I think it's right to flag btw!).

So what needs to be added?

**Comment by WittierDinosaur** [comment]
So there's a couple of things going on when you run this through the parser.

This is the yml output:

```
- statement:
    create_table_statement:
    - keyword: CREATE
    - keyword: TABLE
    - table_reference:
      - identifier: public
      - dot: .
      - identifier: foo
    - bracketed:
      - start_bracket: (
      - column_definition:
          identifier: quadkey
          data_type:
            data_type_identifier: TEXT
      - comma: ','
      - column_definition:
          identifier: geometry
          data_type:
            data_type_identifier: GEOMETRY
            bracketed:
            - start_bracket: (
            - expression:
                column_reference:
                  identifier: GEOMETRY
            - comma: ','
            - expression:
                literal: '4326'
            - end_bracket: )
      - comma: ','
      - table_constraint_definition:
        - keyword: PRIMARY
        - keyword: KEY
        - bracketed:
            start_bracket: (
            column_reference:
              identifier: quadkey
            end_bracket: )
      - end_bracket: )
- statement_terminator: ;
```

It identifies it's a Datatype, but not because GEOMETRY is a keyword. It's because we still have this very liberal Datatype regex:

```
    DatatypeIdentifierSegment=SegmentGenerator(
        # Generate the anti template from the set of reserved keywords
        lambda dialect: RegexParser(
            r"[A-Z][A-Z0-9_]*",
            CodeSegment,
            name="data_type_identifier",
            type="data_type_identifier",
            anti_template=r"^(NOT)$",  # TODO - this is a stopgap until we implement explicit data types
        ),
```

Geometry is definitely not added as a keyword yet.

Secondly, the capitalisation error is because it thinks that the first expression in the brackets "GEOMETRY" is a column, not a Datatype. I don't know whether this issue will fall out if we add it as a keyword, or whether we'll need to explicitly create some valid expressions for PostGIS datatypes, in order for it to look for a keyword inside the brackets.

**Comment by tunetheweb** [comment]
I don't disagree with that, but I would still like confirmation of what the ask is here rather than guess.

The statement parses and lints. Yes possibly 100% not in the correct manner, and possibly that may cause problems later, but it still does parse and lint.

The error that @xamanu has raised is not due to lack of `GEOMETRY` support (though I agree it's not explicitly supported at the moment).

This SQL shows no error, and is valid PostGIS from my reading of that spec (though I've no real knowledge of this so could well be wrong):

```sql
CREATE TABLE public.foo
(
    quadkey TEXT,
    geometry GEOMETRY (geometry, 4326),
    PRIMARY KEY (quadkey)
);
```

You get the same "issue" with this, and I don't think anyone would say we don't support `TABLE`:

```
% echo 'SELECT table FROM TABLE;' | sqlfluff lint -
== [stdin] FAIL
L:   1 | P:  19 | L014 | Unquoted identifiers must be consistently lower case.
All Finished 📜 🎉!
```

So has @xamanu misread the error, or is there a problem that requires fixing? Does `GEOMETRY` need to be in capitals? Do we think it should be exempt from rule L014? Is there other functionality for PostGIS that doesn't lint/parse?

In short: what is the ask here?

**Comment by WittierDinosaur** [comment]
The geometry in brackets is referring to the type GEOMETRY, not the column name. To make the example less confusing:

```
CREATE TABLE public.foo
(
    quadkey TEXT,
    my_geometry_column GEOMETRY (GEOMETRY, 4326),
    PRIMARY KEY (quadkey)
);
```

So the linter is misfiring, expecting us to lowercase the 'identifier' geometry, when it should be a keyword. I've used PostGIS before, and it's creating a geometry column, of type geometry, in geoid id 4326.

**Comment by xamanu** [comment]
Many thanks for looking into this! Yes, I'm happy to confirm that I expected the `GEOMETRY` in the brackets to be in capital letters for the reason that has just been mentioned.

**Comment by xamanu** [comment]
Many thanks!
